### PR TITLE
Fix: correctly order the ground truth and prediction for ARFF files in run.data_content

### PR DESCRIPTION
**Pull request by LennartPurucker**
### Bug - What does this PR fix?
The order of the ground truth and predictions is mixed up in the current implementation of how a run stores prediction data and how it uses the prediction data to build an ARFF file. 

As a result, the ground truth is treated as the predictions, and the predictions are treated as the ground truth. 
Consequently, publishing a run to uploads the wrong values for these columns to the OpenML server. 

### Impact 
This is not validated on the server side (to my understanding). Hence, all ARFF files of predictions uploaded using the OpenML Python API are most likely wrong. Moreover, evaluations of such runs also report the wrong scores. This might have impacted the results of papers that used scores of runs uploaded by the Python Client for meta-analysis. 

#### Reference Issues
Multiple issues exist as a result of this. The following issues are likely related to this problem: #1197, ~~#559~~, #1131

### Fix
I changed the order in the appropiate places and added a test for the (IMO) expected/correct behavior.  Additionally, I changed the tests that checked for the old order to the new order. 

#### New Order 
The order I used follows the order of ARFF files uploaded by the R Client API. I used the following code snippet to find the order. 
```python
from openml.runs import get_run

# 1875133 is a run from mlr 
print(get_run(1875133).predictions_url)

# After downloading and opening the file stored under the predictions_url, one can see that the file's order is:
# @attribute 'repeat' numeric
# @attribute 'fold' numeric
# @attribute 'row_id' numeric
# @attribute 'prediction' {'1','2'}
# @attribute 'truth' {'1','2'}
# @attribute 'confidence.1' numeric
# @attribute 'confidence.2' numeric

``` 

### Open Questions
~~The contribution guidelines mention that I should change the progress.rst. I am unsure if this would be below 13.1 and what I should mention there. What do you think?~~ I updated the progress.rst for the newest version. 

### Required (Server-Side) Follow-up Actions
This bug affects a lot of already published runs on OpenML. 
We might need to change/adjust the uploaded ARFF files and re-evaluate all these runs.

### CODE REVIEW

File: tests/test_runs/test_run.py [hunk 1]
```diff
@@ -189,47 +197,105 @@ def test_to_from_filesystem_no_model(self):
...
+
+        for fold_id in range(s_d[1]):
+            _, test_indices = task.get_train_test_split_indices(repeat=0, fold=fold_id, sample=0)
+            fold_map[test_indices] = fold_id
+
+        X, y = task.get_X_and_y()
+
+        # Check correctness of y_ture and y_pred in run
+        for fold_id in range(s_d[1]):
+            # Get data for fold
```
Comment (mfeurer) [inline tests/test_runs/test_run.py:216]: Why can't we simply read the masks from the splits?

File: tests/test_runs/test_run.py [hunk 2]
```diff
@@ -50,10 +52,16 @@ def _test_run_obj_equals(self, run, run_prime):
                     self.assertDictEqual(other, dict())
         self.assertEqual(run._to_xml(), run_prime._to_xml())
 
-        numeric_part = np.array(np.array(run.data_content)[:, 0:-2], dtype=float)
-        numeric_part_prime = np.array(np.array(run_prime.data_content)[:, 0:-2], dtype=float)
-        string_part = np.array(run.data_content)[:, -2:]
-        string_part_prime = np.array(run_prime.data_content)[:, -2:]
+        # Determine which attributes are numeric and which not
```
[the remark this answers is not in the document]
Comment (LennartPurucker) [reply tests/test_runs/test_run.py:55]: The previous implementation hard-coded for the old order which parts are numeric and which parts are strings. 

This way, we can use the ARFF description to determine the types. If the order changes at some point in the future, this should not brake.

File: tests/test_runs/test_run.py (hunk 1 above)
Comment (LennartPurucker) [reply tests/test_runs/test_run.py:216]: Which function would you use for that / are you referring to?

File: tests/test_runs/test_run.py [hunk 3]
```diff
@@ -189,47 +197,105 @@ def test_to_from_filesystem_no_model(self):
...
-        # obtain run from filesystem
-        loaded_run = openml.runs.OpenMLRun.from_filesystem(cache_path)
-        loaded_run.publish()
-        TestBase._mark_entity_for_removal("run", loaded_run.run_id)
-        TestBase.logger.info(
-            "collected from {}: {}".format(__file__.split("/")[-1], loaded_run.run_id)
-        )
+            # Make sure that the flow has not been uploaded as requested.
+            self.assertFalse(openml.flows.flow_exists(flow.name, flow.external_version))
+            self.assert_run_prediction_data(task, run, clone(model))
```
[the remark this answers is not in the document]
Comment (LennartPurucker) [reply tests/test_runs/test_run.py:286]: Done

File: tests/test_runs/test_run.py (hunk 1 above)
Comment (LennartPurucker) [reply tests/test_runs/test_run.py:246]: Refactored loops to make this better

**Review by mfeurer — APPROVED** [review]
Looks good to me, let's wait for the unit tests to work again and then merge this.

File: doc/progress.rst [hunk 4]
```diff
@@ -9,8 +9,7 @@ Changelog
 0.13.1
 ~~~~~~
 
- * Add new contributions here.
-
+ * FIX #1197 #559 #1131: Fix the order of ground truth and predictions in the ``OpenMLRun`` object and in ``format_prediction``.
```
Comment (PGijsbers) [inline doc/progress.rst:12]: Can you add that it is specifically about regression tasks? thanks

File: openml/runs/functions.py [hunk 5]
```diff
@@ -906,6 +905,14 @@ def obtain_field(xml_obj, fieldname, from_server, cast=None):
 
     tags = openml.utils.extract_xml_tags("oml:tag", run)
 
+    # Make sure default values are used where needed to keep run objects identical
+    if not evaluations:
+        evaluations = None
+    if not fold_evaluations:
+        fold_evaluations = None
+    if not sample_evaluations:
+        sample_evaluations = None
```
Comment (PGijsbers) [inline openml/runs/functions.py:914]: Could instead pass the parameter value `evaluations or None` which means that it's none with an empty dict, and `evaluations` otherwise.

File: doc/progress.rst (hunk 4 above)
Comment (LennartPurucker) [reply doc/progress.rst:12]: The switched order is a problem for regression and classification tasks (and maybe learning curve).

File: openml/runs/functions.py (hunk 5 above)
Comment (LennartPurucker) [reply openml/runs/functions.py]: Indeed, good suggestion, I updated this. Looks much cleaner.

File: tests/test_runs/test_run.py [hunk 6]
```diff
@@ -189,47 +211,155 @@ def test_to_from_filesystem_no_model(self):
...
+            # Assert identical for offline - offline
+            self._test_run_obj_equals(run, loaded_run)
+
+            # Publish and test for offline - online
+            run.publish()
+            self.assertTrue(openml.flows.flow_exists(flow.name, flow.external_version))
+
+            try:
+                online_run = openml.runs.get_run(run.run_id, ignore_cache=True)
+                self._test_run_obj_equals(run, online_run, only_check_prediction_data=True)
```
Comment (PGijsbers) [inline tests/test_runs/test_run.py:356]: Why not instead just call `check_prediction_data` directly?

Comment (LennartPurucker) [reply tests/test_runs/test_run.py:356]: Originally I wanted to check that the object is equal, but this does not work as the runs created from the sever have many differences from locally created runs. Hence, it grew like this. 

But at this point, we could just call `check_prediction_data`. Or do you have any other opinion on this?

Comment (PGijsbers) [reply tests/test_runs/test_run.py:356]: I'm not sure what the differences are.. but if they are not important then just calling `check_prediction_data` should be good enough I think.

File: tests/test_runs/test_run.py [hunk 7]
```diff
@@ -189,47 +211,155 @@ def test_to_from_filesystem_no_model(self):
...
+    def assert_run_prediction_data(task, run, model):
+        # -- Get y_pred and y_true as it should be stored in the run
+        s_d = task.get_split_dimensions()
+        if (s_d[0] > 1) or (s_d[2] > 1):
+            raise ValueError("Test does not support this task type's split dimensions.")
+
+        X, y = task.get_X_and_y()
+
+        # Check correctness of y_ture and y_pred in run
+        for fold_id in range(s_d[1]):
```
Comment (PGijsbers) [inline tests/test_runs/test_run.py:245]:
```suggestion
        n_repeats, n_folds, n_samples = task.get_split_dimensions()
        if (n_repeats > 1) or (n_samples > 1):
            raise ValueError("Test does not support this task type's split dimensions.")

        X, y = task.get_X_and_y()

        # Check correctness of y_true and y_pred in run
        for fold_id in range(n_folds):
```

File: tests/test_runs/test_run.py (hunk 6 above)
Comment (LennartPurucker) [reply tests/test_runs/test_run.py]: Done.

**Review by PGijsbers — APPROVED** [review]

**Review by mfeurer — APPROVED** [review]
